modules/core/CameraThinLens.cpp

```diff
@@ -0,0 +1,33 @@
+#include "CameraThinLens.h"
+#include "Ray.h"
+#include "random.h"
+#include "Sampler.h"
+#include "macroses.h"
+
+namespace rt
+{
+	void CCameraThinLens::InitRay(Ray& ray, int x, int y, const Vec2f& sample)
+	{
+		CCameraPerspective::InitRay(ray, x, y, sample);
+
+		float lensr = getlensRadius();
+		float nBlades = getnBlades();
+		if (lensr > 0) {
+			// Sample point on lens
+			Vec2f pLens;
+			Vec2f sample(random::U<float>(), random::U<float>());
+			
+			//sample from uniformly distributed points in a regular polygon
+			RT_ASSERT(nBlades == 0 || (nBlades >= 3 && nBlades <= 16));
+			pLens = lensr * CSampler::uniformSampleRegularNgon(sample, nBlades, random::u<int>(1, nBlades));
+			
+			// Compute point on plane of focus
+			float ft = getfocalDistance() / ray.dir.val[2];
+			Vec3f pFocus = ray.org + ray.dir * ft;
+
+			// Update ray for effect of lens
+			ray.org += Vec3f(pLens.val[0], pLens.val[1], 0);
```

Comment:
Here you shift the origin in xy-plane, assuming z to be equal to zero. But the lens might lie not only in xy-space. What if the camera looks down or upwards?

Reply:
@ereator yes, I somehow missed this. The orientation of the lens should be same as the camera's orientation instead of just lying on the same plane. I think we can fix this by transforming the sampled points in [-1, 1] so that they lie on the plane formed by the camera's m_xAxis and m_yAxis as follows: 
pLens.val[0] = pLens.val[0]  * m_xAxis;
pLens.val[1] = pLens.val[1] * m_yAxis; 

will test it and commit